src/integrant/core.clj

```diff
@@ -48,12 +48,29 @@
 (defn- sort-keys [ks m]
   (sort (dep/topo-comparator (dependency-graph m)) ks))
 
+(defn- update-key [m k f]
+  (update m k (comp (partial f k) (partial expand-key m))))
+
+(defn build
+  "Apply a function f to each key value pair in a configuration map. Keys are
+  traversed in dependency order, and any references in the value expanded. The
+  function should take two arguments, a key and value, and return a new value."
+  [config keys f]
+  {:pre [(map? config)]}
+  (when-let [refs (seq (missing-refs config))]
+    (throw (ex-info (str "Missing definitions for refs: " (str/join ", " refs))
+                    {:reason ::missing-refs
+                     :config config
+                     :missing-refs refs})))
+  (-> (reduce (fn [m k] (update-key m k f))
+              config
+              (sort-keys keys config))
+      (with-meta {::origin config})))
+
 (defn expand
   "Replace all refs with the values they correspond to."
   [config]
-  (reduce (fn [m k] (update m k (partial expand-key m)))
-          config
-          (sort-keys (keys config) config)))
+  (build config (keys config) (comp second vector)))
```

Comment:
Instead of `(comp second vector)`, which took me a moment to work out, perhaps instead `(fn [_ v] v)`? It's more concise, faster, and easier to see what it does at a glance.